### ISSUE TITLE
JIT Option processing changes

### ISSUE THREAD

**Issue by mpirvu** [eclipse/omr#1055]
**What we are trying to solve:** All features that are implemented in the JIT need options to control their behavior. The options for features that are specific to OMR subprojects should not be included in OMROptions.hpp/cpp (e.g.: options for J9Open features should reside entirely in J9Options.hpp/cpp). Currently this is not possible: any option flags *must* stay in OMROptions.hpp/cpp due to the way options are processed as it will be explained later.

**Goals of this proposal:**
1. Allow subproject options to reside entirely in dedicated option files that are specific to that subproject (e.g. J9Open specific options should only exist in J9Options.hpp/cpp)
1. Use the same method, getOption(optionFlag), to query for option flags no mater where they are defined (in OMROptions.hpp or in J9Options.hpp)

** Details **
For simplicity, in the following I will be referring to J9Open as one of the OMR subprojects. A similar design/implementation can be used for any subproject based on OMR.

Option flags are defined in OMROptions.hpp in an enum, TR_CompilationOptions. The storage for them is an array included in the Options object
````      uint32_t   _options[TR_OWM+1]; ````
If we want to define J9Open specific objects we need a similar enum in J9Options.hpp and add another array of words to store the option flags. Let call this new array _trj9Options[].
To achieve the second goal, the code in getOption(optionFlag) must be able to tell where to look for the flag, either in _options[] or in _trj9Options[]. This can be accomplished by having one of the bits in the optionFlag set for _trj9Options[] and reset for _options[].
````
  #define TR_OWM 0x1f
  #define TR_IDBIT  0x20 // next bit
  
  getOption(uint32_t optionFlag)
     {
     if (optionFlag & TR_IDBIT)
         return (_trj9Options[optionFlag & TR_OWM] & (optionFlag & ~(TR_OWM | TR_IDBIT))) != 0;
     else
        return (_options[optionFlag & TR_OWM] & (optionFlag & ~(TR_OWM | TR_IDBIT))) != 0;
     }
````
If this method can be inlined, then the C++ optimizer should be able to eliminate the 'if' test.
Caveat: if the number of options in OMROptions.hpp exceeds the current capacity, then all the subprojects like trj9 have to change their numbering scheme. Today we have room for 32*(32-5)=864 option flags and we use about 670. With the new approach we will have 832 bits which leaves room for ~160 new options before encoding must change (note that many of the options currently in OMROptions.hpp could/should be moved to J9Options.hpp leaving  even more room to grow)

Note: an alternative to having a new array, _trj9Options[], is to double the size of the exiting _options[] array and impose a restriction that encoding of option flags in OMROptions.hpp cannot exceed half the size of the array.

**The problem with *'base'* **
Options are processed with 
````   OMR::Options::processOption(char *startOption, TR::OptionTable *table,  void *base, int32_t numEntries, TR::OptionSet *optionSet)````
which is called twice for global options:
````
    TR::Options::processOption(options, _jitOptions, jitBase, _numJitEntries, optionSet);
    TR::Options::processOption(options, TR::Options::_feOptions, _feBase, _numVmEntries, optionSet);
````
and for each option set.
*'base'* is instantiated to either jitBase (which is cmdLineOptions) or _feBase (which is jitConfig). 
Options that set flags do so by providing an offset from the 'base'. Thus, for options in J9Open(declared in TR::Options::_feOptions) we actually set flags in jitConfig and there is no way to store flags in the Options object.

**Proposal:**
1) All J9Open options that need to set flags/values in jitConfig will do so through specialized functions that know where to find jitConfig (it is stored in _febase in Options).
2) To process other J9Open options, instead of providing _feBase as a parameter to processOption(), pass the command line options object.
    Similarly, when processing an option subset with  
````feEndOpt = TR::Options::processOption(options, TR::Options::_feOptions, _feBase, _numVmEntries, optionSet);````
       instead of _feBase pass the options object for the subset

**Comment by 0xdaryl** [comment]
@nbhuiyan FYI

**Comment by fjeremic** [comment]
@0xdaryl @nbhuiyan it would be awesome if we could schedule you in for an OMR Compiler Architecture talk on the new framework, just to give us a walk through. You've already written excellent documentation about hte design right in the PR which is much appreciated. Having a conversation about it to a broader audience may be useful as well.

**Comment by nbhuiyan** [comment]
@fjeremic Yes, we were planning on me going over the design of the new option processing framework in the OMR Compiler Architecture meeting scheduled for tomorrow. I'd appreciate getting some feedback from a broader audience about the new design.

**Comment by 0xdaryl** [comment]
That's tomorrow's topic.  Agenda to be sent out.
